tests/ts_cluster_keys_SUITE.erl

```diff
@@ -0,0 +1,722 @@
+%% -------------------------------------------------------------------
+%%
+%% Copyright (c) 2016 Basho Technologies, Inc.
+%%
+%% This file is provided to you under the Apache License,
+%% Version 2.0 (the "License"); you may not use this file
+%% except in compliance with the License.  You may obtain
+%% a copy of the License at
+%%s
+%%   http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing,
+%% software distributed under the License is distributed on an
+%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+%% KIND, either express or implied.  See the License for the
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+%% Tests for the different combinations of keys supported by
+%% Riak Time Series.
+%%
+%% -------------------------------------------------------------------
+-module(ts_cluster_keys_SUITE).
+-compile(export_all).
+
+-include_lib("common_test/include/ct.hrl").
+-include_lib("eunit/include/eunit.hrl").
+
+%%--------------------------------------------------------------------
+%% COMMON TEST CALLBACK FUNCTIONS
+%%--------------------------------------------------------------------
+
+suite() ->
+    [{timetrap,{minutes,10}}].
+
+init_per_suite(Config) ->
+    [Node|_] = Cluster = ts_util:build_cluster(multiple),
+    Pid = rt:pbc(Node),
+    % create tables and populate them with data
+    create_data_def_1(Pid),
+    create_data_def_2(Pid),
+    create_data_def_3(Pid),
+    create_data_def_4(Pid),
+    create_data_def_5(Pid),
+    create_data_def_6(Pid),
+    create_data_def_7(Pid),
+    create_data_def_8(Pid),
+    all_booleans_create_data(Pid),
+    all_timestamps_create_data(Pid),
+    all_types_create_data(Pid),
+    [{cluster, Cluster} | Config].
+
+end_per_suite(_Config) ->
+    ok.
+
+init_per_group(_GroupName, Config) ->
+    Config.
+
+end_per_group(_GroupName, _Config) ->
+    ok.
+
+init_per_testcase(_TestCase, Config) ->
+    Config.
+
+end_per_testcase(_TestCase, _Config) ->
+    ok.
+
+groups() ->
+    [].
+
+all() ->
+    rt:grep_test_functions(?MODULE).
+
+client_pid(Ctx) ->
+    [Node|_] = proplists:get_value(cluster, Ctx),
+    rt:pbc(Node).
+
+%%%
+%%% TABLE 1
+%%%
+
+create_data_def_1(Pid) ->
+    ts_util:assert_row_sets({[],[]},riakc_ts:query(Pid, table_def_1())),
+    ok = riakc_ts:put(Pid, <<"table1">>, [[1,1,N,1] || N <- lists:seq(1,6000)]).
+
+column_names_def_1() ->
+    [<<"a">>, <<"b">>, <<"c">>, <<"d">>].
+
+table_def_1() ->
+    "CREATE TABLE table1 ("
+    "a SINT64 NOT NULL, "
+    "b SINT64 NOT NULL, "
+    "c TIMESTAMP NOT NULL, "
+    "d SINT64 NOT NULL, "
+    "PRIMARY KEY  ((a,b,quantum(c, 1, 's')), a,b,c,d))".
+
+select_exclusive_def_1_test(Ctx) ->
+    Query =
+        "SELECT * FROM table1 WHERE a = 1 AND b = 1 AND c > 0 AND c < 11",
```

Comment:
I would consider creating a function to return these query strings to make sure the arguments for the results and the query string always are in sync:
```
query_range_string(Start, End, StartInclusive, EndInclusive) ->
    io_lib:format("SELECT * FROM table1 WHERE a = 1 AND b = 1 AND c ~p ~B AND c ~p ~B",
                         lists:flatten([start_incl(StartInclusive, Start), end_incl(EndInclusive, End+1)]).

start_incl(true, Start) -> [">=", Start];
start_incl(false, Start) -> [">", Start-1].

end_incl(true, End) -> ["<=", End];
end_incl(false, End) -> ["<", End+1].
```

And the same for the results:
```
results_list(Start, End) ->
    [{1, 1, N, 1} || N <- lists:seq(Start, End)].
```

Which allows us to write a generic test function:
```
generic_select_test_args(Start, End, StartIncl, EndIncl, Ctx) ->
    Query = query_range_string(Start, End, StartIncl, EndIncl),
    Results = results_list(Start, End),
    {{column_names_def_1(), Results},
       riakc_ts:query(client_pid(Ctx), Query)}
```

And then a test can be written like this:
```
select_exclusive_def_1_test(Ctx) ->
    {Exp, Res} = generic_select_test_args(1,10, false, false, Ctx),
    ts_util:assert_row_sets(Exp, Res).
```

The final comparison has to be in the test to ease finding the problem in my experience, but there could be smarter ways.

Not necessary to it exactly as I have written - just the main idea of making the query string and the results align based on arguments instead of knowing it. The best way of doing that is the desired outcome.